Renderer/Renderer/RayCasting/Geometry/PrimitiveObject.h

```diff
@@ -0,0 +1,57 @@
+#pragma once
+
+#include <Renderer/RayCasting/Geometry/SceneGeometry.h>
+
+#include <Physics/Primitives/Disc.h>
+#include <Physics/Primitives/Plane.h>
+#include <Physics/Primitives/Ray.h>
+#include <Physics/Primitives/Sphere.h>
+#include <Physics/Primitives/Triangle.h>
+
+namespace GLEngine::Renderer {
+
+template <class T, typename = decltype(std::declval<T>().IntersectImpl(std::declval<const Physics::Primitives::S_Ray&>()))> class C_Primitive : public I_RayGeometryObject {
+public:
+	C_Primitive(T primitive)
+		: m_Primitive(primitive)
+		, m_Area(RayTracing::T_GeometryTraits::GetArea(primitive))
+	{
+	}
+
+	[[nodiscard]] virtual bool Intersect(const Physics::Primitives::S_Ray& ray, C_RayIntersection& intersection) const override
+	{
+		const auto t = m_Primitive.IntersectImpl(ray);
+		if (t <= 0)
+			return false;
+
+		FillIntersection(m_Primitive, t, ray, intersection);
+		intersection.SetMaterial(&GetMaterial());
+		return true;
+	}
+	[[nodiscard]] virtual float Area() const override { return m_Area; }
+
+	T m_Primitive;
+
+private:
+	void FillIntersection(const Physics::Primitives::S_Plane& plane, float t, const Physics::Primitives::S_Ray& ray, C_RayIntersection& intersection) const
```

Comment:
I can move this into the cpp or traits file. It looks horrible here in the header.